compiler/sem/semtypes.nim

```diff
@@ -232,12 +228,32 @@ proc semVarOutType(c: PContext, n: PNode, prev: PType; kind: TTypeKind): PType =
   else:
     result = newConstraint(c, kind)
 
+proc isRecursiveType(t: PType, cycleDetector: var IntSet): bool =
+  if t.isNil:
+    return false
+  if cycleDetector.containsOrIncl(t.id):
+    true
+  else:
+    case t.kind
+    of tyAlias, tyGenericInst, tyDistinct:
+      isRecursiveType(t.lastSon, cycleDetector)
+    else:
+      false
+
 proc semDistinct(c: PContext, n: PNode, prev: PType): PType =
   if n.len == 0: return newConstraint(c, tyDistinct)
   result = newOrPrevType(tyDistinct, prev, c)
-  addSonSkipIntLit(result, semTypeNode(c, n[0], nil), c.idgen)
+  let t = semTypeNode(c, n[0], nil).skipIntLit(c.idgen)
+  var cycleDetector = initIntSet()
+  if isRecursiveType(t, cycleDetector):
+    result.add t
+    c.config.localReport(n.info, reportTyp(rsemIllegalRecursion, t))
+  else:
+    rawAddSon(result, t)
```

Comment:
```suggestion
  let t = semTypeNode(c, n[0], nil).skipIntLit(c.idgen)
  if prev != nil and isReachable(prev, t):
    result = newOrPrevType(tyError, prev, c)
    c.config.localReport(n.info, reportTyp(rsemIllegalRecursion, t))
  else:
    result = newOrPrevType(tyDistinct, prev, c)
    rawAddSon(result, t)
```
Assuming the introduction of the `isReachable` from above. The earlier `result` assignment needs to be removed for this to work. Creating a `tyError` makes sure that no cyclic type is introduced in the case of `nimsuggest` and `nim check`.

Reply:
Updated